import os
import subprocess
import configparser
import argparse
class Myconf(configparser.ConfigParser):
    def __init__(self, defaults=None):
        configparser.ConfigParser.__init__(self, defaults=defaults)

    def optionxform(self, optionstr):
        return optionstr

def run(target_bed,probe_bed,bam,outdir,prefix,configfile):
    config = Myconf()
    config.read(configfile)
    java = config.get('software', 'java')
    picard = config.get('software', 'picard2.20.6')
    gatk4 = config.get('software', 'gatk4.1.3')
    gatk3 = config.get('software', 'gatk3.7')
    bedtools=config.get('software','bedtools2.28.0')
    ref=config.get('database', 'hg19_ref')
    if not os.path.exists(outdir):
        os.mkdir(outdir)
    out=outdir+"/"+prefix
    ####BedToIntervalList (Picard)#############
    cmd="%s -jar %s BedToIntervalList I=%s O=%s/target.interval_list SD=%s" %(java,picard,target_bed,outdir,bam)
    subprocess.check_call(cmd,shell=True)
    cmd = "%s -jar %s BedToIntervalList I=%s O=%s/probe.interval_list SD=%s" % (java,picard, probe_bed, outdir, bam)
    subprocess.check_call(cmd, shell=True)
    ####Metrics generated by CollectHsMetrics for the analysis of target-capture sequencing experiments#######
    cmd="%s -Xmx40g -jar %s CollectHsMetrics I=%s O=%s.hs_metrics.txt R=%s TARGET_INTERVALS=%s/target.interval_list" \
        " BAIT_INTERVALS=%s/probe.interval_list COVMAX=1000000 "%(java,picard,bam,out,ref,outdir,outdir)
    subprocess.check_call(cmd,shell=True)
    ################InsertSize####################
    cmd="%s -Xmx40g -jar %s CollectInsertSizeMetrics I=%s O=%s.insert_size_metrics.txt H=%s.insert_size_histogram.pdf"%(java,picard,bam,out,out)
    subprocess.check_call(cmd,shell=True)
    ####Assess sequence coverage by a wide array of metrics, partitioned by sample, read group, or library#############
    cmd = "%s -Xmx40g -jar %s -T DepthOfCoverage -allowPotentiallyMisencodedQuals --minBaseQuality 20 --minMappingQuality 20 -R %s -I %s -o %s " \
          " -ct 1 -ct 10 -ct 20 -ct 50 -ct 100 " \
          " -ct 150 -ct 200 -ct 250 -ct 500 -L %s/target.interval_list --omitDepthOutputAtEachBase --omitIntervalStatistics" \
          % (java,gatk3, ref, bam, out, outdir)
    subprocess.check_call(cmd, shell=True)
    ################Count On-Target Reads##########
    cmd = os.popen("%s intersect -bed -u -abam %s -b %s | wc -l"%(bedtools,bam,target_bed))
    reads = cmd.read()
    reads = reads.strip()
    outfile = open("%s.MetricsReport.tsv" % (out), "w")
    outfile.write("Counts-On-Target_Reads\t%s\n"%(reads))
    ################################################
    infile = open("%s.insert_size_metrics.txt" % (out), "r")
    num, name, counts = 0, [], 0
    for line in infile:
        line = line.strip()
        array = line.split("\t")
        counts += 1
        if line.startswith("MEDIAN_INSERT_SIZE"):
            num = counts
            for i in range(len(array)):
                name.append(array[i])
        if counts == num + 1 and num != 0:
            print(line)
            outfile.write("meadian_insert_size\t%s\n" % (array[0]))
    infile.close()
    ###############################################
    infile=open("%s.hs_metrics.txt"%(out),"r")
    num=0
    name=[]
    for line in infile:
        if not line.startswith("#"):
            array=line.split("\t")
            if array[0]=="BAIT_SET":
                for i in range(len(array)):
                    name.append(array[i])
                num += 1
            else:
                if num==1 and len(array)==len(name):
                    for i in range(len(array)):
                        if name[i] == "FOLD_80_BASE_PENALTY":
                            outfile.write("FOLD_80_BASE_PENALTY\t%s\n"%(array[i]))
                        if name[i] == "MEAN_TARGET_COVERAGE":
                            outfile.write("mean_depth\t%s\n" % (array[i]))
                        if name[i] == "MEDIAN_TARGET_COVERAGE":
                            outfile.write("median_depth\t%s\n" % (array[i]))
    infile.close()
    ###############################################
    infile = open("%s.sample_summary" % (out), "r")
    num, name = 0, []
    for line in infile:
        num += 1
        line = line.strip()
        array = line.split("\t")
        if num == 1:
            for i in range(len(array)):
                name.append(array[i])
        if num == 2:
            for i in range(len(array)):
                if name[i] == "%_bases_above_1":
                    outfile.write("%%_bases_above_1X\t%s\n" % (array[i]))
                if name[i] == "%_bases_above_20":
                    outfile.write("%%_bases_above_20X\t%s\n" % (array[i]))
                if name[i] == "%_bases_above_50":
                    outfile.write("%%_bases_above_50X\t%s\n" % (array[i]))
                if name[i] == "%_bases_above_100":
                    outfile.write("%%_bases_above_100X\t%s\n" % (array[i]))
                if name[i] == "%_bases_above_200":
                    outfile.write("%%_bases_above_200X\t%s\n" % (array[i]))
                if name[i] == "%_bases_above_250":
                    outfile.write("%%_bases_above_250X\t%s\n" % (array[i]))
                if name[i] == "%_bases_above_500":
                    outfile.write("%%_bases_above_500X\t%s\n" % (array[i]))
    infile.close()
    outfile.close()
    outfile.close()
    subprocess.check_call("rm -rf %s.sample*"%(out),shell=True)
if __name__=="__main__":
    parser=argparse.ArgumentParser("Bam stat.")
    parser.add_argument("--target",help="target bed",required=True)
    parser.add_argument("--probe", help="probe bed(or target bed)", required=True)
    parser.add_argument("--config",help="config file",required=True)
    parser.add_argument("--bam",help="input bam file must remove or mark PCR duplicates",required=True)
    parser.add_argument("--outdir",help="output directory",required=True)
    parser.add_argument("--prefix",help="prefix of output",required=True)
    args=parser.parse_args()
    run(args.target, args.probe, args.bam, args.outdir, args.prefix,args.config)